src/hardware/mixer.cpp

```diff
@@ -1367,35 +1388,81 @@ void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 
 	last_samples_were_stereo = stereo;
 
-	auto &convert_out = do_resampler ? mixer.resample_temp : mixer.resample_out;
+	auto &convert_out = do_resample ? mixer.resample_temp : mixer.resample_out;
 	ConvertSamples<Type, stereo, signeddata, nativeorder>(data, frames, convert_out);
 
-	auto out_frames = check_cast<spx_uint32_t>(convert_out.size()) / 2u;
+	auto out_frames = check_cast<uint32_t>(convert_out.size()) / 2u;
+
+	if (do_resample) {
+		switch (resample_method) {
+		case ResampleMethod::LinearInterpolation: {
+			auto &s = lerp_upsampler;
+
+			auto in_pos  = mixer.resample_temp.begin();
+			auto &out = mixer.resample_out;
+			out.resize(0);
 
-	if (do_resampler) {
-		auto in_frames = check_cast<spx_uint32_t>(mixer.resample_temp.size()) / 2u;
+			AudioFrame curr_frame = {*in_pos, *(in_pos + 1)};
 
-		out_frames = estimate_max_out_frames(resampler.state, in_frames);
+			out_frames = 0;
 
-		mixer.resample_out.resize(out_frames * 2);
+			while (in_pos != mixer.resample_temp.end()) {
+				const auto out_left = lerp(s.last_frame.left,
+				                           curr_frame.left,
+				                           s.pos);
 
-		speex_resampler_process_interleaved_float(resampler.state,
-		                                          mixer.resample_temp.data(),
-		                                          &in_frames,
-		                                          mixer.resample_out.data(),
-		                                          &out_frames);
+				const auto out_right = lerp(s.last_frame.right,
+				                            curr_frame.right,
+				                            s.pos);
 
-		// out_frames now contains the actual number of resampled frames,
-		// ensure the number of output frames is within the logical size.
-		assert(out_frames <= mixer.resample_out.size() / 2);
-		mixer.resample_out.resize(out_frames * 2);  // only shrinks
+				out.emplace_back(out_left);
+				out.emplace_back(out_right);
+				++out_frames;
+
+				s.pos += s.step;
+
+				if (s.pos > 1.0f) {
+					s.pos -= 1.0f;
+					// Move to the next input frame
+					in_pos += 2;
+					s.last_frame = curr_frame;
+					curr_frame   = {*in_pos, *(in_pos + 1)};
+				}
+			}
```

Comment:
When we get to the end, `curr_frame` on line 1431 will dereference the `end()` and `end() + 1`.

here's a simple walk through with a two-frame vector:

Take `resample_temp[4] = { left1, right1, left2, right2, [end==invalid] }`

1. `in_pos` starts on `left1` and `curr_frame = {left1, right1}`
2. line 1427 bumps `in_pos` by two: now on `left2`
3. `curr_frame = {left2, right2}`
4. while loop passes again, we're not on `end()` yet.
5. line 1427 bumps `in_pos` by another two, which now on `end()`
6. `s.last_frame =` is correctly assigned to the prior `{left2, right2}`
7. But now `curr_frame = {*end(), *end() + 1}`

Reply:
Hmm, you're right; I'll move loading the current frame to the top.

Reply:
Indeed, the it crashes the sanitizer build right away... Should remember to always do a final test with it.

Reply:
(good to know the sanitizer caught it too!)